In [1]:
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import pandas as pd
import seaborn as sns
import warnings

In [2]:
warnings.filterwarnings('ignore')

In [18]:
dataset = pd.read_csv('Melbourne_housing_FULL.csv')
print(dataset.info())
dataset.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34857 entries, 0 to 34856
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         34857 non-null  object 
 1   Address        34857 non-null  object 
 2   Rooms          34857 non-null  int64  
 3   Type           34857 non-null  object 
 4   Price          27247 non-null  float64
 5   Method         34857 non-null  object 
 6   SellerG        34857 non-null  object 
 7   Date           34857 non-null  object 
 8   Distance       34856 non-null  float64
 9   Postcode       34856 non-null  float64
 10  Bedroom2       26640 non-null  float64
 11  Bathroom       26631 non-null  float64
 12  Car            26129 non-null  float64
 13  Landsize       23047 non-null  float64
 14  BuildingArea   13742 non-null  float64
 15  YearBuilt      15551 non-null  float64
 16  CouncilArea    34854 non-null  object 
 17  Lattitude      26881 non-null  float64
 18  Longti

Suburb             Address  Rooms Type      Price Method  SellerG  \
0  Abbotsford       68 Studley St      2    h        NaN     SS   Jellis   
1  Abbotsford        85 Turner St      2    h  1480000.0      S   Biggin   
2  Abbotsford     25 Bloomburg St      2    h  1035000.0      S   Biggin   
3  Abbotsford  18/659 Victoria St      3    u        NaN     VB   Rounds   
4  Abbotsford        5 Charles St      3    h  1465000.0     SP   Biggin   
5  Abbotsford    40 Federation La      3    h   850000.0     PI   Biggin   
6  Abbotsford         55a Park St      4    h  1600000.0     VB   Nelson   
7  Abbotsford        16 Maugie St      4    h        NaN     SN   Nelson   
8  Abbotsford        53 Turner St      2    h        NaN      S   Biggin   
9  Abbotsford        99 Turner St      2    h        NaN      S  Collins   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  3/09/2016       2.5    3067.0  ...       1.0  1.0     126.0           NaN   
1  3/12/2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
2  4/02/2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
3  4/02/2016       2.5    3067.0  ...       2.0  1.0       0.0           NaN   
4  4/03/2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   
5  4/03/2017       2.5    3067.0  ...       2.0  1.0      94.0           NaN   
6  4/06/2016       2.5    3067.0  ...       1.0  2.0     120.0         142.0   
7  6/08/2016       2.5    3067.0  ...       2.0  2.0     400.0         220.0   
8  6/08/2016       2.5    3067.0  ...       1.0  2.0     201.0           NaN   
9  6/08/2016       2.5    3067.0  ...       2.0  1.0     202.0           NaN   

   YearBuilt         CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN  Yarra City Council  -37.8014    144.9958  Northern Metropolitan   
1        NaN  Yarra City Council  -37.7996    144.9984  Northern Metropolitan   
2     1900.0  Yarra City Council  -37.8079    144.9934  Northern Metropolitan   
3        NaN  Yarra City Council  -37.8114    145.0116  Northern Metropolitan   
4     1900.0  Yarra City Council  -37.8093    144.9944  Northern Metropolitan   
5        NaN  Yarra City Council  -37.7969    144.9969  Northern Metropolitan   
6     2014.0  Yarra City Council  -37.8072    144.9941  Northern Metropolitan   
7     2006.0  Yarra City Council  -37.7965    144.9965  Northern Metropolitan   
8     1900.0  Yarra City Council  -37.7995    144.9974  Northern Metropolitan   
9     1900.0  Yarra City Council  -37.7996    144.9989  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  
5        4019.0  
6        4019.0  
7        4019.0  
8        4019.0  
9        4019.0  

[10 rows x 21 columns]

In [4]:
dataset.nunique()

Suburb             351
Address          34009
Rooms               12
Type                 3
Price             2871
Method               9
SellerG            388
Date                78
Distance           215
Postcode           211
Bedroom2            15
Bathroom            11
Car                 15
Landsize          1684
BuildingArea       740
YearBuilt          160
CouncilArea         33
Lattitude        13402
Longtitude       14524
Regionname           8
Propertycount      342
dtype: int64

In [5]:
# let's use limited columns which makes more sense for serving our purpose
cols_to_use = ['Suburb', 'Rooms', 'Type', 'Method', 'SellerG', 'Regionname', 'Propertycount', 
               'Distance', 'CouncilArea', 'Bedroom2', 'Bathroom', 'Car', 'Landsize', 'BuildingArea', 'Price']
dataset = dataset[cols_to_use]
print(dataset.shape)
dataset.head()

(34857, 15)


Suburb  Rooms Type Method SellerG             Regionname  \
0  Abbotsford      2    h     SS  Jellis  Northern Metropolitan   
1  Abbotsford      2    h      S  Biggin  Northern Metropolitan   
2  Abbotsford      2    h      S  Biggin  Northern Metropolitan   
3  Abbotsford      3    u     VB  Rounds  Northern Metropolitan   
4  Abbotsford      3    h     SP  Biggin  Northern Metropolitan   

   Propertycount  Distance         CouncilArea  Bedroom2  Bathroom  Car  \
0         4019.0       2.5  Yarra City Council       2.0       1.0  1.0   
1         4019.0       2.5  Yarra City Council       2.0       1.0  1.0   
2         4019.0       2.5  Yarra City Council       2.0       1.0  0.0   
3         4019.0       2.5  Yarra City Council       3.0       2.0  1.0   
4         4019.0       2.5  Yarra City Council       3.0       2.0  0.0   

   Landsize  BuildingArea      Price  
0     126.0           NaN        NaN  
1     202.0           NaN  1480000.0  
2     156.0          79.0  1035000.0  
3       0.0           NaN        NaN  
4     134.0         150.0  1465000.0

In [6]:
dataset.isna().sum()

Suburb               0
Rooms                0
Type                 0
Method               0
SellerG              0
Regionname           3
Propertycount        3
Distance             1
CouncilArea          3
Bedroom2          8217
Bathroom          8226
Car               8728
Landsize         11810
BuildingArea     21115
Price             7610
dtype: int64

In [7]:
cols_to_fill_zero = ['Propertycount', 'Distance', 'Bedroom2', 'Bathroom', 'Car']
dataset[cols_to_fill_zero] = dataset[cols_to_fill_zero].fillna(0)

dataset['Landsize'] = dataset['Landsize'].fillna(dataset.Landsize.mean())
dataset['BuildingArea'] = dataset['BuildingArea'].fillna(dataset.BuildingArea.mean())

dataset.isna().sum()

Suburb              0
Rooms               0
Type                0
Method              0
SellerG             0
Regionname          3
Propertycount       0
Distance            0
CouncilArea         3
Bedroom2            0
Bathroom            0
Car                 0
Landsize            0
BuildingArea        0
Price            7610
dtype: int64

In [23]:
new = dataset[dataset.Price.isna()]
new.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7610 entries, 0 to 34850
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         7610 non-null   object 
 1   Address        7610 non-null   object 
 2   Rooms          7610 non-null   int64  
 3   Type           7610 non-null   object 
 4   Price          0 non-null      float64
 5   Method         7610 non-null   object 
 6   SellerG        7610 non-null   object 
 7   Date           7610 non-null   object 
 8   Distance       7610 non-null   float64
 9   Postcode       7610 non-null   float64
 10  Bedroom2       5834 non-null   float64
 11  Bathroom       5831 non-null   float64
 12  Car            5706 non-null   float64
 13  Landsize       5065 non-null   float64
 14  BuildingArea   3086 non-null   float64
 15  YearBuilt      3467 non-null   float64
 16  CouncilArea    7610 non-null   object 
 17  Lattitude      5888 non-null   float64
 18  Longtit

In [25]:
cols_to_fill_zero = ['Propertycount', 'Distance', 'Bedroom2', 'Bathroom', 'Car']
new[cols_to_fill_zero] = new[cols_to_fill_zero].fillna(0)

new['Landsize'] = new['Landsize'].fillna(new.Landsize.mean())
new['BuildingArea'] = new['BuildingArea'].fillna(new.BuildingArea.mean())

new.isna().sum() 

Suburb              0
Address             0
Rooms               0
Type                0
Price            7610
Method              0
SellerG             0
Date                0
Distance            0
Postcode            0
Bedroom2            0
Bathroom            0
Car                 0
Landsize            0
BuildingArea        0
YearBuilt        4143
CouncilArea         0
Lattitude        1722
Longtitude       1722
Regionname          0
Propertycount       0
dtype: int64

In [26]:
new = new[cols_to_use]

In [30]:
n_inp = new.drop('Price', axis=1)
n_tar = new['Price']

n_inp.isna().sum() 

Suburb           0
Rooms            0
Type             0
Method           0
SellerG          0
Regionname       0
Propertycount    0
Distance         0
CouncilArea      0
Bedroom2         0
Bathroom         0
Car              0
Landsize         0
BuildingArea     0
dtype: int64

In [34]:
n_inp = pd.get_dummies(n_inp, drop_first=True)
print(n_inp.info())
n_inp.head()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7610 entries, 0 to 34850
Columns: 625 entries, Rooms to CouncilArea_Yarra Ranges Shire Council
dtypes: float64(7), int64(1), uint8(617)
memory usage: 5.0 MB
None


Rooms  Propertycount  Distance  Bedroom2  Bathroom  Car  Landsize  \
0      2         4019.0       2.5       2.0       1.0  1.0     126.0   
3      3         4019.0       2.5       3.0       2.0  1.0       0.0   
7      4         4019.0       2.5       3.0       2.0  2.0     400.0   
8      2         4019.0       2.5       4.0       1.0  2.0     201.0   
9      2         4019.0       2.5       3.0       2.0  1.0     202.0   

   BuildingArea  Suburb_Aberfeldie  Suburb_Airport West  ...  \
0    172.071973                  0                    0  ...   
3    172.071973                  0                    0  ...   
7    220.000000                  0                    0  ...   
8    172.071973                  0                    0  ...   
9    172.071973                  0                    0  ...   

   CouncilArea_Moorabool Shire Council  CouncilArea_Moreland City Council  \
0                                    0                                  0   
3                                    0                                  0   
7                                    0                                  0   
8                                    0                                  0   
9                                    0                                  0   

   CouncilArea_Nillumbik Shire Council  CouncilArea_Port Phillip City Council  \
0                                    0                                      0   
3                                    0                                      0   
7                                    0                                      0   
8                                    0                                      0   
9                                    0                                      0   

   CouncilArea_Stonnington City Council  CouncilArea_Whitehorse City Council  \
0                                     0                                    0   
3                                     0                                    0   
7                                     0                                    0   
8                                     0                                    0   
9                                     0                                    0   

   CouncilArea_Whittlesea City Council  CouncilArea_Wyndham City Council  \
0                                    0                                 0   
3                                    0                                 0   
7                                    0                                 0   
8                                    0                                 0   
9                                    0                                 0   

   CouncilArea_Yarra City Council  CouncilArea_Yarra Ranges Shire Council  
0                               1                                       0  
3                               1                                       0  
7                               1                                       0  
8                               1                                       0  
9                               1                                       0  

[5 rows x 625 columns]

In [8]:
dataset.dropna(inplace=True)
dataset.shape

(27244, 15)

In [33]:
dataset = pd.get_dummies(dataset, drop_first=True)
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34857 entries, 0 to 34856
Columns: 34884 entries, Rooms to Regionname_Western Victoria
dtypes: float64(12), int64(1), uint8(34871)
memory usage: 1.1 GB


In [10]:
x = dataset.drop('Price', axis=1)
y = dataset['Price']

In [11]:
from sklearn.model_selection import train_test_split
train_x, test_x, train_y, test_y = train_test_split(x, y, test_size=0.3, random_state=2)

In [12]:
from sklearn.linear_model import LinearRegression
reg = LinearRegression().fit(train_x, train_y)

In [13]:
print(reg.score(test_x, test_y))
print(reg.score(train_x, train_y))

0.13853683161537644
0.6827792395792723


Here training score is 68% but test score is 13.85% which is very low

Normal Regression is clearly overfitting the data, let's try other models

### Using Lasso (L1 Regularized) Regression Model

In [14]:
from sklearn import linear_model
l1_reg = linear_model.Lasso(alpha=50, max_iter=100, tol=0.1)
l1_reg.fit(train_x, train_y)

print(l1_reg.score(test_x, test_y))
print(l1_reg.score(train_x, train_y))

0.6636111369404489
0.6766985624766824


#### Using Ridge (L2 Regularized) Regression Model

In [15]:
from sklearn.linear_model import Ridge
l2_reg = Ridge(alpha=50, max_iter=100, tol=0.1)
l2_reg.fit(train_x, train_y)

print(l2_reg.score(test_x, test_y))
print(l2_reg.score(train_x, train_y))

0.6670848945194958
0.6622376739684328


**We see that Lasso and Ridge Regularizations prove to be beneficial when our Simple Linear Regression Model overfits. These results may not be that contrast but significant in most cases.Also that L1 & L2 Regularizations are used in Neural Networks too**

In [16]:
y_pre = l1_reg.predict(test_x)
new = pd.DataFrame(y_pre)
new['2'] = l2_reg.predict(test_x)

In [17]:
new.describe()

0             2
count  8.174000e+03  8.174000e+03
mean   1.051919e+06  1.052720e+06
std    5.362933e+05  5.128310e+05
min   -2.384868e+06 -2.217022e+05
25%    6.798758e+05  6.919819e+05
50%    9.865999e+05  9.944497e+05
75%    1.363246e+06  1.355469e+06
max    4.349892e+06  3.125603e+06

In [32]:
n_tar = l2_reg.predict(n_inp)

ValueError: X has 625 features, but Ridge is expecting 744 features as input.